src/block/Hopper.php

```diff
@@ -92,8 +99,231 @@ public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player
 	}
 
 	public function onScheduledUpdate() : void{
-		//TODO
+		$this->position->getWorld()->scheduleDelayedBlockUpdate($this->position, 1);
+
+		$tile = $this->position->getWorld()->getTile($this->position);
+		if(!$tile instanceof TileHopper){
+			return;
+		}
+
+		$transferCooldown = $tile->getTransferCooldown();
+		if($transferCooldown > 0){
+			$transferCooldown--;
+			$tile->setTransferCooldown($transferCooldown);
+		}
+
+		if($this->isPowered() || $transferCooldown > 0){
+			return;
+		}
+
+		$inventory = $tile->getInventory();
+		$success = $this->push($inventory);
+		// Hoppers that have a container above them, won't try to pick up items.
+		$origin = $this->position->getWorld()->getTile($this->position->getSide(Facing::UP));
+		//TODO: Not all blocks a hopper can pull from have an inventory (for example: Jukebox).
+		if($origin instanceof Container){
+			$success = $this->pull($inventory, $origin->getInventory()) || $success;
+		}else{
+			$success = $this->pickup($inventory) || $success;
+		}
+		// The cooldown is only set back to the default amount of ticks if the hopper has done anything.
+		if($success){
+			$tile->setTransferCooldown(TileHopper::DEFAULT_TRANSFER_COOLDOWN);
+		}
+	}
+
+	/**
+	 * This function handles pushing items from the hopper to a tile in the direction the hopper is facing.
+	 * Returns true if an item was successfully pushed or false on failure.
+	 */
+	private function push(HopperInventory $inventory) : bool{
+		if(count($inventory->getContents()) === 0){
+			return false;
+		}
+		$destination = $this->position->getWorld()->getTile($this->position->getSide($this->facing));
+		if($destination === null){
+			return false;
+		}
+
+		for($slot = 0; $slot < $inventory->getSize(); $slot++){
+			$item = $inventory->getItem($slot);
+			if($item->isNull()){
+				continue;
+			}
+
+			// Hoppers interact differently when pushing into different kinds of tiles.
+			//TODO: Composter
+			//TODO: Brewing Stand
+			//TODO: Jukebox (improve)
+			if($destination instanceof \pocketmine\block\tile\Furnace){
```

Comment:
Each push logic should be implemented in the different tiles to allow greater flexibility / addition in the future.

Reply:
Better implement it in the blocks, not just to remove the dependence on tiles, but also don't force yourself only to allow the implementation of hopper logic for tiles: There could be blocks, that hoppers can interact with, without them being a tile. (I don't know the Bedrock internals well enough, so I don't know if composters would also need a tile. But if not, they would be a counterexample.)

Reply:
That's indirectly what I was explaining, but you did it better than I did. I mentioned tile because that's what was checked, but implementing it in blocks seems much more coherent.